pymc_marketing/mmm/budget_optimizer.py

```diff
@@ -13,217 +13,185 @@
-        The List of channels for which the budget is being optimized.
-    parameters : Dict
-        Michaelis-Menten parameters for each channel as described in `calculate_expected_contribution`.
-
-    Returns
-    -------
-    float
-        Negative of the total expected contribution for the given budget distribution.
+    adstock : AdstockTransformation
+        The adstock class.
+    saturation : SaturationTransformation
+        The saturation class.
+    num_days : int
+        The number of days.
+    parameters : dict
+        A dictionary of parameters for each channel.
+    adstock_first : bool, optional
+        Whether to apply adstock transformation first or saturation transformation first.
+        Default is True.
     """
 
-    sum_contributions = 0.0
-
-    for channel, budget in zip(channels, x, strict=False):
-        if method == "michaelis-menten":
-            L, k = parameters[channel]
-            sum_contributions += michaelis_menten(budget, L, k)
-
-        elif method == "sigmoid":
-            alpha, lam = parameters[channel]
-            sum_contributions += sigmoid_saturation(budget, alpha, lam)
-
+    def __init__(
+        self,
+        adstock: AdstockTransformation,
+        saturation: SaturationTransformation,
+        num_days: int,
+        parameters: dict[str, dict[str, dict[str, float]]],
+        adstock_first: bool = True,
+    ):
+        self.adstock = adstock
+        self.saturation = saturation
+        self.num_days = num_days
+        self.parameters = parameters
+        self.adstock_first = adstock_first
+
+    def objective(self, budgets: list[float]) -> float:
+        """
+        Calculate the total response during a period of time given the budgets,
+        considering the saturation and adstock transformations.
+
+        Parameters
+        ----------
+        budgets : list[float]
+            The budgets for each channel.
+
+        Returns
+        -------
+        float
+            The negative total response value.
+        """
+        total_response = 0
+        first_transform, second_transform = (
+            (self.adstock, self.saturation)
+            if self.adstock_first
+            else (self.saturation, self.adstock)
+        )
+        for idx, (_channel, params) in enumerate(self.parameters.items()):
+            budget = budgets[idx]
+            first_params = (
+                params["adstock_params"]
+                if self.adstock_first
+                else params["saturation_params"]
+            )
+            second_params = (
+                params["saturation_params"]
+                if self.adstock_first
+                else params["adstock_params"]
+            )
+            spend = np.full(self.num_days, budget)
+            spend_extended = np.concatenate([spend, np.zeros(self.adstock.l_max)])
+            transformed_spend = second_transform.function(
+                x=first_transform.function(x=spend_extended, **first_params).eval(),
```

Comment:
do we need this inner eval?
```suggestion
                x=first_transform.function(x=spend_extended, **first_params),
```